app/Http/Controllers/Auth/LoginController.php

```diff
@@ -37,4 +41,32 @@ public function __construct()
     {
         $this->middleware('guest')->except('logout');
     }
+
+    public function socialLogin($social)
+    {
+        return Socialite::driver($social)->redirect();
+    }
+
+    public function socialRedirect($social)
+    {
+        $user = Socialite::driver($social)->user();
+        
+        if(isset($user->name)){
+            $name = $user->name;
+        }else{
+            $name = $user->nickname;
+        }
+
+        $user = User::firstOrCreate([
```

Comment:
Giải thích logic ý tưởng đoạn này e nhé.

Reply:
hàm firstOrCreate() sẽ kiểm tra tất cả các đối số truyền vào. Nếu không có kết quả nào hoàn toàn trùng khớp thì nó sẽ tạo 1 cái mảng user mới. Hàm ở trên em truyền vào đối số là email = $user->email.